relay-quotas/src/global.rs

```diff
@@ -0,0 +1,231 @@
+use std::sync::{Arc, Mutex, OnceLock, PoisonError};
+
+use relay_common::time::UnixTimestamp;
+use relay_redis::redis::Script;
+use relay_redis::{PooledClient, RedisError};
+
+use crate::RedisQuota;
+
+fn load_global_lua_script() -> &'static Script {
+    static SCRIPT: OnceLock<Script> = OnceLock::new();
+    SCRIPT.get_or_init(|| Script::new(include_str!("global_quota.lua")))
+}
+
+fn current_slot(window: u64) -> usize {
+    UnixTimestamp::now()
+        .as_secs()
+        .checked_div(window)
+        .unwrap_or_default() as usize
+}
+
+/// A rate limiter for global rate limits.
+#[derive(Default)]
+pub struct GlobalRateLimits {
+    limits: Mutex<hashbrown::HashMap<BudgetKey, Arc<Mutex<GlobalRateLimit>>>>,
+}
+
+impl GlobalRateLimits {
+    /// Returns `true` if the global quota should be ratelimited.
+    ///
+    /// Certain errors can be resolved by syncing to redis, so in those cases
+    /// we try again to decrement the budget after syncing.
+    pub fn is_rate_limited(
+        &self,
+        client: &mut PooledClient,
+        quota: &RedisQuota,
+        quantity: usize,
+    ) -> Result<bool, RedisError> {
+        let Some(limit) = quota.limit else {
+            return Ok(false);
+        };
+
+        let key = BudgetKeyRef::new(quota);
+        let val = {
+            let mut limits = self.limits.lock().unwrap_or_else(PoisonError::into_inner);
```

Comment:
Yeah that is fine, we never actually panic in the critical section, but even if we do, there is nothing that is left in an unsafe or uncertain state, since all we do is either insert something into the map or clone an Arc.